src/Microsoft.OData.Core/ODataUtils.cs

```diff
@@ -206,11 +206,11 @@ public static ODataServiceDocument GenerateServiceDocument(this IEdmModel model)
 
             ODataServiceDocument serviceDocument = new ODataServiceDocument();
             serviceDocument.EntitySets = model.EntityContainer.EntitySets()
-                .Select(entitySet => new ODataEntitySetInfo() { Name = entitySet.Name, Title = entitySet.Name, Url = new Uri(entitySet.Name, UriKind.RelativeOrAbsolute) }).ToList();
+                .Select(entitySet => new ODataEntitySetInfo() { Name = entitySet.Name, Title = entitySet.Name, Url = new Uri(entitySet.Name, UriKind.RelativeOrAbsolute) });

```

Comment:
Each call to .EntitySets(), .Singletons(), .OperationImports, etc. calls through EntityContainer.Elements(),OfType<> for each element type.  Better would be to remove the LINQ all together -- build 3 lists and iterate through model.EntityContainer.Elements(), adding the appropriate element to each list.